### PR TITLE
Make global alloc hugepages failure flag atomic

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim please review

### CODE REVIEW

File: vespalib/src/vespa/vespalib/util/alloc.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
 #include "alloc.h"
+#include "atomic.h"
```
Comment (baldersheim) [inline vespalib/src/vespa/vespalib/util/alloc.cpp:3]: <atomic> ?

File: vespalib/src/vespa/vespalib/util/alloc.cpp [hunk 2]
```diff
@@ -328,8 +331,8 @@ MMapAllocator::salloc(size_t sz, void * wantedAddress)
         }
         buf = mmap(wantedAddress, sz, prot, flags | _G_HugeFlags, -1, 0);
         if (buf == MAP_FAILED) {
-            if ( ! _G_hasHugePageFailureJustHappened ) {
-                _G_hasHugePageFailureJustHappened = true;
+            if ( ! load_relaxed(_G_hasHugePageFailureJustHappened)) {
```
Comment (baldersheim) [inline vespalib/src/vespa/vespalib/util/alloc.cpp:334]: if ( ! _G_hasHugePageFailureJustHappened.exchange(true)) {  atomic load/and store

File: vespalib/src/vespa/vespalib/util/alloc.cpp [hunk 3]
```diff
@@ -347,8 +350,8 @@ MMapAllocator::salloc(size_t sz, void * wantedAddress)
                 }
             }
         } else {
-            if (_G_hasHugePageFailureJustHappened) {
-                _G_hasHugePageFailureJustHappened = false;
+            if (load_relaxed(_G_hasHugePageFailureJustHappened)) {
```
Comment (baldersheim) [inline vespalib/src/vespa/vespalib/util/alloc.cpp:353]: if ( _G_hasHugePageFailureJustHappened.exchange(false)) { atomic load/and store

**Review by baldersheim — CHANGES_REQUESTED** [review]

File: vespalib/src/vespa/vespalib/util/alloc.cpp (hunk 2 above)
Comment (vekterli) [reply vespalib/src/vespa/vespalib/util/alloc.cpp:334]: Is it worth a `cmpxchg` to skip a debug log?

Comment (baldersheim) [reply vespalib/src/vespa/vespalib/util/alloc.cpp:334]: No, good point.

File: vespalib/src/vespa/vespalib/util/alloc.cpp (hunk 1 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/util/alloc.cpp:3]: Ah, was not aware of those.

File: vespalib/src/vespa/vespalib/util/alloc.cpp (hunk 3 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/util/alloc.cpp:353]: good point.

**Review by baldersheim — APPROVED** [review]

**Pull request merged**